std/socket.d

```diff
@@ -1269,6 +1269,14 @@ abstract class Address
     /// Returns actual size of underlying $(D sockaddr) structure.
     abstract @property socklen_t nameLen() const pure nothrow @nogc;
 
+    // Sets the actual size of underlying $(D sockaddr).
+    // Kludge for setting path length in UnixAddress
+    protected @property void nameLen(socklen_t len) pure
+    {
+        if (len > this.nameLen)
+            throw new SocketParameterException("Not enough socket address storage");
+    }
+
```

Comment:
The reasoning behind adding this setter was:
There are methods in `Socket` (`remoteAddress`, `localAddress`, `receiveFrom`), which do the following:
1. call `Socket.createAddress`, getting an instance of one of `Address`' descendants, created with argument-less constructor.
2. call `Address.name` and `Address.nameLen` getters on that instance to get the address and capacity of the buffer for the next call
3. call one of `getsockname`, `getpeername`, `recvfrom`, to set the name in the buffer just obtained.
4. read `Address.nameLen` again and compare with result of previous call to see if the buffer was large enough (`receiveFrom` doesn't do this step)

While the above procedure works well for fixed-length addresses, it doesn't provide a way to feed the actual length of the address back to the concrete `Address`. I had to add a way that doesn't change the public interface nor the behavior of the above methods for fixed-length addresses, and does not require adding a special case for `UnixAddress` in all of them.

The `nameLen` setter is protected and deliberately not documented with DDoc. The default implementation in base `Address` is intended to be sufficient for the typical case: fixed-length addresses.
Yes, there is one case where it breaks the expectation you mentioned: where the `len` argument passed is less than the fixed-length. I should add a check for this.


Another solution I had in mind was to remove the argument-less constructors, get the address from `getsockname`, `getpeername` and `receiveFrom` to a temporary buffer on the stack, and refactor `createAddress` to create concrete `Address`es by passing that temporary buffer to the appropriate constructor.